Common/Render/Text/draw_text_sdl.cpp

```diff
@@ -0,0 +1,476 @@
+	const int ptSize = (int)((size + 6) / dpiScale_);
+
+	TTF_Font *font = TTF_OpenFont(useFont, ptSize);
+
+	if (!font) {
+		File::FileInfo fileInfo;
+   		g_VFS.GetFileInfo("Roboto-Condensed.ttf", &fileInfo);
+		font = TTF_OpenFont(fileInfo.fullName.c_str(), ptSize);
+	}
+
+	fontMap_[fontHash] = font;
+	fontHash_ = fontHash;
+	return fontHash;
+}
+
+void TextDrawerSDL::SetFont(uint32_t fontHandle) {
+	uint32_t fontHash = fontHandle;
+	auto iter = fontMap_.find(fontHash);
+	if (iter != fontMap_.end()) {
+		fontHash_ = fontHandle;
+	} else {
+		ERROR_LOG(G3D, "Invalid font handle %08x", fontHandle);
+	}
+}
+
+void TextDrawerSDL::MeasureString(const char *str, size_t len, float *w, float *h) {
+	CacheKey key{ std::string(str, len), fontHash_ };
+
+	TextMeasureEntry *entry;
+	auto iter = sizeCache_.find(key);
+	if (iter != sizeCache_.end()) {
+		entry = iter->second.get();
+	} else {
+		TTF_Font *font = fontMap_.find(fontHash_)->second;
+		int ptSize = TTF_FontHeight(font) / 1.35;
+
+		uint32_t missingGlyph = CheckMissingGlyph(key.text);
+		
+		if (missingGlyph && FindFallbackFonts(missingGlyph, ptSize)) {
+			font = fallbackFonts_[0];
+		}
+
+		int width = 0;
+		int height = 0;
+		TTF_SizeUTF8(font, key.text.c_str(), &width, &height);
+
+		entry = new TextMeasureEntry();
+		entry->width = width;
+		entry->height = height;
+		sizeCache_[key] = std::unique_ptr<TextMeasureEntry>(entry);
+	}
+
+	entry->lastUsedFrame = frameCount_;
+	*w = entry->width * fontScaleX_ * dpiScale_;
+	*h = entry->height * fontScaleY_ * dpiScale_;
+}
+
+void TextDrawerSDL::MeasureStringRect(const char *str, size_t len, const Bounds &bounds, float *w, float *h, int align) {
+	std::string toMeasure = std::string(str, len);
+	int wrap = align & (FLAG_WRAP_TEXT | FLAG_ELLIPSIZE_TEXT);
+	if (wrap) {
+		bool rotated = (align & (ROTATE_90DEG_LEFT | ROTATE_90DEG_RIGHT)) != 0;
+		WrapString(toMeasure, toMeasure.c_str(), rotated ? bounds.h : bounds.w, wrap);
+	}
+
+	TTF_Font *font = fontMap_.find(fontHash_)->second;
+	int ptSize = TTF_FontHeight(font) / 1.35;
+	uint32_t missingGlyph = CheckMissingGlyph(toMeasure);
+
+	if (missingGlyph && FindFallbackFonts(missingGlyph, ptSize)) {
+		font = fallbackFonts_[0];
+	}
+
+	std::vector<std::string> lines;
+	SplitString(toMeasure, '\n', lines);
+
+	int total_w = 0;
+	int total_h = 0;
+	for (size_t i = 0; i < lines.size(); i++) {
+		CacheKey key{ lines[i], fontHash_ };
+
+		TextMeasureEntry *entry;
+		auto iter = sizeCache_.find(key);
+		if (iter != sizeCache_.end()) {
+			entry = iter->second.get();
+		} else {
+			int width = 0;
+			int height = 0;
+			TTF_SizeUTF8(font, lines[i].c_str(), &width, &height);
+			entry = new TextMeasureEntry();
+			entry->width = width;
+			entry->height = height;
+			sizeCache_[key] = std::unique_ptr<TextMeasureEntry>(entry);
+		}
+
+		entry->lastUsedFrame = frameCount_;
+		if (total_w < entry->width) {
+			total_w = entry->width;
+		}
+		total_h += entry->height;
```

Comment:
This should probably be TTF_LineSkip() or something.  But maybe we can just pass the string with newlines to TTF_SizeUTF8(), it'll give correct height?  I'm not sure.  Just noting that fonts don't always have the same height as line spacing.

Especially since you're using Wrapped to render (which is probably a good thing), this could cause multiline panels to be have awkward spacing below them (assuming the text rendering func respects line spacing.)

-[Unknown]

Reply:
TTF_SizeUTF8() doesn't handle wrapped text unfortunately.